### ISSUE TITLE
inconsistencies resolving .^ in an app vs a generator

### ISSUE THREAD

**Issue by joemfb** [urbit/arvo#191]
Not sure if this is the right place for this ... 

I'm butting my head trying to resolve a `.^` inside an app, when the same twig works in a generator.

In `:dojo`:

```
> .^(json %cx /=home/348/web/easy/json)
>> [%camping %c %x [p=~wolsyp-filtuc-havrut-londys--pinbel-ballut-fodmyl-litmer q=%home r=[%ud p=348]] s=/json/easy/web]
... :: json object printed
```

When I evaluate that twig in a generator, I see the same `[%camping ...]` printf, but inside an app, it just prints the path and blocks:

```
/cx/~wolsyp-filtuc-havrut-londys--pinbel-ballut-fodmyl-litmer/home/348/web/easy/json
```

(same behavior in the `++prep` and `++poke` arms).

While I have ~1000 other questions about `.^`, which is pretty likely documented, this is the obstacle I haven't been able to work around.

**Comment by joemfb** [comment]
Seemingly relevant tidbit:

In `:dojo`:

```
> .^(hart %e /(scot %p our)/host/real)
[p=%.y q=~ r=[%.y p=<|org urbit wolsyp-filtuc-havrut-londys--pinbel-ballut-fodmyl-litmer|>]]
```

Whereas in `:talk`, the same twig just produces the path:

```
/e/~wolsyp-filtuc-havrut-londys--pinbel-ballut-fodmyl-litmer/host/real
```

**Comment by cgyarvin** [comment]
Wise question, glasshoppa!

The answer: blocking .^ is a very non-fundamental behavior.  It is
obviously not part of Nock proper (it turns into Nock instruction 11, which
is only implemented by the virtual extended Nock ++mock).  It is not even
available to %gall apps, which can only do non-blocking .^.  For the
programmer, the first layer blocking .^ works in is anything run by :dojo
(generators, command line, etc).

Internally, what you're doing with .^ from the dojo, and what you're doing
from a %gall app, are different.  From a %gall app, .^ will just use the
Arvo kernel namespace function ++scry on your behalf.  ++scry obviously has
no way to block.  So in Unix terms, it's essentially producing EWOULDBLOCK.

From the dojo, .^ is actually invoking the build vane, %ford.  The
superpowers of %ford are tremendous (and not well documented).  You're
essentially saying: start this build and call me back when it's done.  If
%ford does not actually have to block, it will give a result in the same
top-level Arvo step.  But it's quite prepared to block.

(Moreover, the way we implement "blocking" is to... abort the computation
and rerun it when we have the result we need.  With proper caching this is
surprisingly not-bad.)

If you want to do the same thing in :talk or another app, you need to
invoke %ford asynchronously the way :dojo does.  Sorry, there is no such
thing as "real" app blocking.  I can defend this: one of the things about
Urbit's super-event-centric I/O model is that your app state is always just
a noun, never some kind of saved execution context.  The illusion of
blocking is pushed up to the highest possible level.  The result of this is
that it makes sense to, for example, hotpatch a blocked :dojo.  There is no
"please exit the program to continue upgrading" or whatever.

**Comment by cgyarvin** [comment]
Yeah, except what he's scrying *exists*, and shouldn't fail in any way.

Yours,
r

**Comment by cgyarvin** [comment]
He's scrying into %eyre, so maybe that's a question for Anton!

**Comment by cgyarvin** [comment]
The original example was a clay scry.

Yours,
r

**Comment by ohAitch** [comment]
By "in talk", do you mean in a talk eval, or in the talk app?

In the talk app, I would expect the eyre scry to work, but not the clay
one: clay scry into the past is also blocking due to marks needing to be
rewound.

**Comment by joemfb** [comment]
I mean a `:talk` eval... I just added that example since it's a similar
test-case, and demonstrates the same dichotomy.


Thanks all for the details! My most pressing follow-up: how do I resolve a
non-blocking :wish? Is there some type of explicit callback mechanism, or
do I have to produce (and then respond to) moves?

**Comment by cgyarvin** [comment]
In practice, the flow for an event request versus a ++scry is totally
different.  So if a specific request succeeds on one path rather than
another, the place to look is that path.  They *should* match...

**Comment by joemfb** [comment]
Yep, you gotta do it with moves.

**Comment by joemfb** [comment]
I don't know what happened on that email thread, but your response became "my" comment here ... ;)

My original twig is `.^(json %cx /=home/348/web/easy/json)`; what kind of move does that translate into?

**Comment by juped** [comment]
Sounds like a bug!

**Comment by ohAitch** [comment]
Ford turns it into a clay subscription, %warp. It's probably easiest to
send the request to ford as a twig; you'll have to unwrap the vase somehow,
but that's the case either way.

**Comment by joemfb** [comment]
This all makes sense, conceptually, but I'm having some trouble making it concrete  Can you point towards any examples of that pattern?

In more detail, a `++move` is a `{bone term wire *}`, at least as taught in the Arvo tutorial (elsewhere, it's a bunch of other things ;)).

Here's a basic recursive move in an app -- it pokes the current app with an atom of 1. (`ost`, `our`, and `dap` are in the subject via `bowl`):

`[[ost %poke /foobar [our dap] %atom 1] ~]`

If I'm reading that correctly `ost` is the `bone`, `%poke` is the term, and `[/foobar [our dap]` is the wire, while `%atom` is references a mark and `1` is the arbitrary argument to our move.

What should I change to create a %clay %warp subscription, ie, what should the term and wire be?

**Comment by joemfb** [comment]
Something like [ost %warp /foobar [our our] %home `[%sing %x ud+134
/web/easy/json]]~, looking at lib/kiln.hoon and arvo/zuse.hoon

**Comment by ohAitch** [comment]
Hm, the email reply authors seem to be getting somewhat scrambled.

**Comment by joemfb** [comment]
It's a creative take on pseudonymity ...

**Comment by philipcmonk** [comment]
Github is always right, y'all are just crazy.

**Comment by philipcmonk** [comment]
In your example, `/foobar` is the wire, and `[[our dap] %atom 1]` is the `*` part of the move.

For a clay subscription, `%warp` is your term, and the wire can be anything you want.  The rest of the move should be a `sock` and a `riff`, as you can see by searching for `$warp` in arvo/zuse.hoon.  `sock` is a pair of our ship and the other ship (also our ship in this case), and `riff` is the request itself.

**Comment by joemfb** [comment]
I get `[%swim-lost %c %warp]` from your example. I'll dig into zuse and kiln and try to pull this apart. Thanks!

**Comment by joemfb** [comment]
I've been meaning to investigate this further, so I haven't commented back... I never did get the file-load moves to work (trying only with raw moves ... `++foal` and company are probably easier).

However, the advice here is wrong.

You *can* use `.^` in an app, you just can't use the beak placeholders `/===/`. You have to use an explicit, "fully qualified path". See `:talk`: https://github.com/urbit/arvo/blob/master/app/talk.hoon#L2317.

ie, this common pattern:

```
/(scot %p our)/<explicit-desk>/(scot %da now)/...
```

**Comment by philipcmonk** [comment]
I suspect that .^ will indeed work in an app as long as it doesn't have to block for any reason, just like `++sky` functions in vanes.  For clay, that usually means local ship, current version.
